framework/test/gtest/ServiceFactoryTest.cpp

```diff
@@ -0,0 +1,169 @@
+        return MakeInterfaceMap<ITestServiceA>(implPtr);
+      }));
+  EXPECT_CALL(*sf, UngetService(testing::_, testing::_, testing::_)).Times(0);
+
+  ServiceRegistration<ITestServiceA, ITestServiceB> reg1 =
+    context.RegisterService<ITestServiceA, ITestServiceB>(
+      ToFactory(sf),
+      { { Constants::SERVICE_SCOPE, Any(Constants::SCOPE_PROTOTYPE) } });
+
+  auto sref1 = context.GetServiceReference<ITestServiceA>();
+  ASSERT_TRUE(static_cast<bool>(sref1));
+  FrameworkEvent lastEvent;
+  auto lToken = context.AddFrameworkListener(
+    [](const cppmicroservices::FrameworkEvent& evt) {
+      ASSERT_EQ(evt.GetType(), FrameworkEvent::FRAMEWORK_WARNING);
+    });
+  ASSERT_EQ(context.GetService<ITestServiceA>(sref1), nullptr);
+  auto sref2 = context.GetServiceReference<ITestServiceB>();
+  ASSERT_TRUE(static_cast<bool>(sref2));
+  ASSERT_EQ(context.GetService<ITestServiceB>(sref2), nullptr);
+  context.RemoveListener(std::move(lToken));
+}
+
+// Tests the return type and FrameworkEvent generated when a
+// ServiceFactory instance throws an exception in it's GetService
+// callback
+TEST_F(ServiceFactoryTest, TestGetServiceThrows)
+{
+  auto context = framework.GetBundleContext();
+  auto sf = std::make_shared<MockFactory>();
+  EXPECT_CALL(*sf, GetService(testing::_, testing::_))
+    .Times(1)
+    .WillRepeatedly(testing::Throw(std::runtime_error("Some exception")));
+  EXPECT_CALL(*sf, UngetService(testing::_, testing::_, testing::_)).Times(0);
+
+  ServiceRegistration<ITestServiceA> reg1 =
+    context.RegisterService<ITestServiceA>(
+      ToFactory(sf),
+      { { Constants::SERVICE_SCOPE, Any(Constants::SCOPE_PROTOTYPE) } });
+
+  auto sref = context.GetServiceReference<ITestServiceA>();
+  ASSERT_TRUE(static_cast<bool>(sref));
+  auto lToken = context.AddFrameworkListener(
+    [](const cppmicroservices::FrameworkEvent& evt) {
+      ASSERT_EQ(evt.GetType(), FrameworkEvent::FRAMEWORK_ERROR);
+    });
+  ASSERT_EQ(context.GetService<ITestServiceA>(sref), nullptr);
+  context.RemoveListener(std::move(lToken));
+}
+
+// Tests the return type and FrameworkEvent generated when a
+// ServiceFactory instance returns nullptr
+TEST_F(ServiceFactoryTest, TestGetServiceReturnsNull)
+{
+  auto context = framework.GetBundleContext();
+  auto sf = std::make_shared<MockFactory>();
+  EXPECT_CALL(*sf, GetService(testing::_, testing::_))
+    .Times(1)
+    .WillRepeatedly(testing::Return(nullptr));
+  EXPECT_CALL(*sf, UngetService(testing::_, testing::_, testing::_)).Times(0);
+
+  ServiceRegistration<ITestServiceA> reg1 =
+    context.RegisterService<ITestServiceA>(
+      ToFactory(sf),
+      { { Constants::SERVICE_SCOPE, Any(Constants::SCOPE_PROTOTYPE) } });
+
+  auto sref = context.GetServiceReference<ITestServiceA>();
+  ASSERT_TRUE(static_cast<bool>(sref));
+  auto lToken = context.AddFrameworkListener(
+    [](const cppmicroservices::FrameworkEvent& evt) {
+      ASSERT_EQ(evt.GetType(), FrameworkEvent::FRAMEWORK_WARNING);
```

Comment:
I believe that the ```FrameworkEvent``` will have an exception in it. The test should assert that the exception is in the event.